cmd/ecr-credential-provider/main.go

```diff
@@ -59,40 +69,112 @@ func defaultECRProvider(region string, registryID string) (*ecr.ECR, error) {
 	return ecr.New(sess), nil
 }
 
-func (e *ecrPlugin) GetCredentials(ctx context.Context, image string, args []string) (*v1.CredentialProviderResponse, error) {
-	registryID, region, registry, err := parseRepoURL(image)
+func publicECRProvider() (*ecrpublic.ECRPublic, error) {
+	// ECR public registries are only in one region and only accessible from regions
+	// in the "aws" partition.
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config:            aws.Config{Region: aws.String(ecrPublicRegion)},
+		SharedConfigState: session.SharedConfigEnable,
+	})
 	if err != nil {
 		return nil, err
 	}
 
+	return ecrpublic.New(sess), nil
+}
+
+type credsData struct {
+	registry  string
+	authToken *string
+	expiresAt *time.Time
+}
+
+func (e *ecrPlugin) getPublicCredsData() (credsData, error) {
+	klog.Infof("Getting creds for public registry")
+	var err error
+
+	if e.ecrPublic == nil {
+		e.ecrPublic, err = publicECRProvider()
+	}
+	if err != nil {
+		return credsData{}, err
+	}
+
+	output, err := e.ecrPublic.GetAuthorizationToken(&ecrpublic.GetAuthorizationTokenInput{})
+	if err != nil {
+		return credsData{}, err
+	}
+
+	if output == nil {
+		return credsData{}, errors.New("response output from ECR was nil")
+	}
+
+	if output.AuthorizationData == nil {
+		return credsData{}, errors.New("authorization data was empty")
+	}
+
+	return credsData{
+		registry:  ecrPublicURL,
+		authToken: output.AuthorizationData.AuthorizationToken,
+		expiresAt: output.AuthorizationData.ExpiresAt,
+	}, nil
+}
+
+func (e *ecrPlugin) getPrivateCredsData(image string) (credsData, error) {
```

Comment:
Same as above. Could we switch to the following signature to return `nil` instead of empty data structure? `credsData{}`

```
func (e *ecrPlugin) getPublicCredsData() (*credsData, error)
```